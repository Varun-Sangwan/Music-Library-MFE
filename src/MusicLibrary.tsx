import React, { useMemo, useState } from "react";
import songs from "./data/songs";
import { GroupBy, SortBy, UserSelectionState } from "./MusicLibrary.types";
import { songsListToDisplay } from "./MusicLibrary.logic";
import "./index.css";

const MusicLibrary: React.FC = () => {
  const [userSelections, setUserSelections] = useState<UserSelectionState>({
    filter: "",
    sortBy: "title",
    groupBy: "none",
  });

  const songsToDisplay = useMemo(
    () => songsListToDisplay(songs, userSelections),
    [userSelections]
  );

  console.log({ songsToDisplay });
  return (
    <div className="music-library-container">
      <h1 className="music-library-title">Music Library</h1>
      <input
        type="text"
        placeholder="Search by title, artist, album..."
        value={userSelections.filter}
        onChange={(e) =>
          setUserSelections({ ...userSelections, filter: e.target.value })
        }
        data-testid="filter-input"
        className="search-bar"
      />
      <div className="filter-controls">
        <select
          onChange={(e) =>
            setUserSelections({
              ...userSelections,
              sortBy: e.target.value as SortBy,
            })
          }
          data-testid="sort-select"
        >
          <option value="title">Sort by Title</option>
          <option value="artist">Sort by Artist</option>
          <option value="album">Sort by Album</option>
        </select>

        <select
          onChange={(e) =>
            setUserSelections({
              ...userSelections,
              groupBy: e.target.value as GroupBy,
            })
          }
          data-testid="group-select"
        >
          <option value="none">No Grouping</option>
          <option value="artist">Group by Artist</option>
          <option value="album">Group by Album</option>
        </select>
      </div>
      {songsToDisplay?.All ? (
        <div className="song-list">
          {songsToDisplay?.All.map((song) => (
            <div key={song.id} className="song-card">
              <img src={song.image} alt={song.album} className="song-image" />
              <p className="song-title">{song.title}</p>
              <p className="song-details">
                {song.artist} ({song.album})
              </p>
            </div>
          ))}
        </div>
      ) : (
        Object.entries(songsToDisplay).map(([group, songs]) => (
          <>
            <h2 className="song-group">{group}</h2>
            <div className={"song-list"}>
              {songs.map((song) => (
                <div key={song.id} className="song-card">
                  <img
                    src={song.image}
                    alt={song.album}
                    className="song-image"
                  />
                  <p className="song-title">{song.title}</p>
                  <p className="song-details">
                    {song.artist} ({song.album})
                  </p>
                </div>
              ))}
            </div>
          </>
        ))
      )}
    </div>
  );
};

export default MusicLibrary;
