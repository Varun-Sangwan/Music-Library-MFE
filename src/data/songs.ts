import { SongsType } from "../MusicLibrary.types";

const songs: SongsType[] = [
  {
    id: 1,
    title: "Bohemian Rhapsody",
    artist: "Queen",
    album: "A Night at the Opera",
    image:
      "https://upload.wikimedia.org/wikipedia/en/4/42/Beatles_-_Abbey_Road.jpg",
  },
  {
    id: 2,
    title: "Love of My Life",
    artist: "Queen",
    album: "A Night at the Opera",
    image:
      "https://upload.wikimedia.org/wikipedia/en/4/42/Beatles_-_Abbey_Road.jpg",
  },
  {
    id: 3,
    title: "You're My Best Friend",
    artist: "Queen",
    album: "A Night at the Opera",
    image:
      "https://upload.wikimedia.org/wikipedia/en/4/42/Beatles_-_Abbey_Road.jpg",
  },
  {
    id: 4,
    title: "Stairway to Heaven",
    artist: "Led Zeppelin",
    album: "Led Zeppelin IV",
    image:
      "https://upload.wikimedia.org/wikipedia/en/2/26/Led_Zeppelin_-_Led_Zeppelin_IV.jpg",
  },
  {
    id: 5,
    title: "Black Dog",
    artist: "Led Zeppelin",
    album: "Led Zeppelin IV",
    image:
      "https://upload.wikimedia.org/wikipedia/en/2/26/Led_Zeppelin_-_Led_Zeppelin_IV.jpg",
  },
  {
    id: 6,
    title: "Rock and Roll",
    artist: "Led Zeppelin",
    album: "Led Zeppelin IV",
    image:
      "https://upload.wikimedia.org/wikipedia/en/2/26/Led_Zeppelin_-_Led_Zeppelin_IV.jpg",
  },
  {
    id: 7,
    title: "Hotel California",
    artist: "Eagles",
    album: "Hotel California",
    image: "https://upload.wikimedia.org/wikipedia/en/4/49/Hotelcalifornia.jpg",
  },
  {
    id: 8,
    title: "New Kid in Town",
    artist: "Eagles",
    album: "Hotel California",
    image: "https://upload.wikimedia.org/wikipedia/en/4/49/Hotelcalifornia.jpg",
  },
  {
    id: 9,
    title: "Life in the Fast Lane",
    artist: "Eagles",
    album: "Hotel California",
    image: "https://upload.wikimedia.org/wikipedia/en/4/49/Hotelcalifornia.jpg",
  },
  {
    id: 10,
    title: "Imagine",
    artist: "John Lennon",
    album: "Imagine",
    image: "https://upload.wikimedia.org/wikipedia/en/4/49/Hotelcalifornia.jpg",
  },
  {
    id: 11,
    title: "Hey Jude",
    artist: "The Beatles",
    album: "The Beatles",
    image: "https://upload.wikimedia.org/wikipedia/en/4/49/Hotelcalifornia.jpg",
  },
  {
    id: 12,
    title: "Come Together",
    artist: "The Beatles",
    album: "Abbey Road",
    image:
      "https://upload.wikimedia.org/wikipedia/en/4/42/Beatles_-_Abbey_Road.jpg",
  },
  {
    id: 13,
    title: "Something",
    artist: "The Beatles",
    album: "Abbey Road",
    image:
      "https://upload.wikimedia.org/wikipedia/en/4/42/Beatles_-_Abbey_Road.jpg",
  },
  {
    id: 14,
    title: "Billie Jean",
    artist: "Michael Jackson",
    album: "Thriller",
    image:
      "https://upload.wikimedia.org/wikipedia/en/5/55/Michael_Jackson_-_Thriller.png",
  },
  {
    id: 15,
    title: "Thriller",
    artist: "Michael Jackson",
    album: "Thriller",
    image:
      "https://upload.wikimedia.org/wikipedia/en/5/55/Michael_Jackson_-_Thriller.png",
  },
  {
    id: 16,
    title: "Beat It",
    artist: "Michael Jackson",
    album: "Thriller",
    image:
      "https://upload.wikimedia.org/wikipedia/en/5/55/Michael_Jackson_-_Thriller.png",
  },
  {
    id: 17,
    title: "Smells Like Teen Spirit",
    artist: "Nirvana",
    album: "Nevermind",
    image:
      "https://upload.wikimedia.org/wikipedia/en/b/b7/NirvanaNevermindalbumcover.jpg",
  },
  {
    id: 18,
    title: "Come as You Are",
    artist: "Nirvana",
    album: "Nevermind",
    image:
      "https://upload.wikimedia.org/wikipedia/en/b/b7/NirvanaNevermindalbumcover.jpg",
  },
  {
    id: 19,
    title: "Rolling in the Deep",
    artist: "Adele",
    album: "21",
    image: "https://upload.wikimedia.org/wikipedia/en/1/1b/Adele_-_21.png",
  },
  {
    id: 20,
    title: "Someone Like You",
    artist: "Adele",
    album: "21",
    image: "https://upload.wikimedia.org/wikipedia/en/1/1b/Adele_-_21.png",
  },
  {
    id: 21,
    title: "Lose Yourself",
    artist: "Eminem",
    album: "8 Mile Soundtrack",
    image: "https://upload.wikimedia.org/wikipedia/en/4/49/Hotelcalifornia.jpg",
  },
  {
    id: 22,
    title: "Without Me",
    artist: "Eminem",
    album: "The Eminem Show",
    image: "https://upload.wikimedia.org/wikipedia/en/3/35/The_Eminem_Show.jpg",
  },
  {
    id: 23,
    title: "Blinding Lights",
    artist: "The Weeknd",
    album: "After Hours",
    image:
      "https://upload.wikimedia.org/wikipedia/en/4/42/Beatles_-_Abbey_Road.jpg",
  },
  {
    id: 24,
    title: "Save Your Tears",
    artist: "The Weeknd",
    album: "After Hours",
    image:
      "https://upload.wikimedia.org/wikipedia/en/4/42/Beatles_-_Abbey_Road.jpg",
  },
];

export default songs;
