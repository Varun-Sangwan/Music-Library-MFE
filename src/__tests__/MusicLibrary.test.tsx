import { render, screen, fireEvent } from "@testing-library/react";
import MusicLibrary from "../MusicLibrary";
import songs from "../data/songs";
import { songsListToDisplay } from "../MusicLibrary.logic";
import { vi } from "vitest";

vi.mock("../MusicLibrary.logic", () => ({
  songsListToDisplay: vi.fn(),
}));

describe("MusicLibrary Component", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("renders the Music Library title", () => {
    (songsListToDisplay as jest.Mock).mockReturnValue({ All: songs });

    render(<MusicLibrary />);

    expect(screen.getByText("Music Library")).toBeInTheDocument();
  });

  it("updates filter when typing in the search input", () => {
    (songsListToDisplay as jest.Mock).mockReturnValue({ All: songs });

    render(<MusicLibrary />);

    const searchInput = screen.getByPlaceholderText(
      "Search by title, artist, album..."
    );

    fireEvent.change(searchInput, { target: { value: "Beatles" } });

    expect(searchInput).toHaveValue("Beatles");
  });

  it("displays filtered and sorted songs when selections change", () => {
    const filteredSongs = {
      All: [{ id: 1, title: "Song A", artist: "Artist X", album: "Album 1" }],
    };

    (songsListToDisplay as jest.Mock).mockReturnValue(filteredSongs);

    render(<MusicLibrary />);

    expect(screen.getByText("Song A")).toBeInTheDocument();
    expect(screen.queryByText("Song B")).not.toBeInTheDocument();
  });

  it("displays grouped songs correctly", () => {
    const groupedSongs = {
      "Album 1": [
        { id: 1, title: "Song A", artist: "Artist X", album: "Album 1" },
        { id: 2, title: "Song C", artist: "Artist Y", album: "Album 1" },
      ],
    };

    (songsListToDisplay as jest.Mock).mockReturnValue(groupedSongs);

    render(<MusicLibrary />);

    expect(screen.getByText("Album 1")).toBeInTheDocument();
    expect(screen.getByText("Song A")).toBeInTheDocument();
    expect(screen.getByText("Song C")).toBeInTheDocument();
  });

  it("updates sorting when selecting an option", () => {
    (songsListToDisplay as jest.Mock).mockReturnValue({ All: [] });

    render(<MusicLibrary />);

    const sortSelect = screen.getByTestId("sort-select");
    fireEvent.change(sortSelect, { target: { value: "artist" } });

    expect(sortSelect).toHaveValue("artist");
  });

  it("updates grouping when selecting an option", () => {
    (songsListToDisplay as jest.Mock).mockReturnValue({ All: [] });

    render(<MusicLibrary />);

    const groupSelect = screen.getByTestId("group-select");
    fireEvent.change(groupSelect, { target: { value: "album" } });

    expect(groupSelect).toHaveValue("album");
  });
});
