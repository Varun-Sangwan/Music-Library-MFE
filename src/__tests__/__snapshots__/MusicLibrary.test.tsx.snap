// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MusicLibrary Component > renders App component correctly 1`] = `
<DocumentFragment>
  <div
    class="p-6 bg-gray-200 text-white min-h-screen"
  >
    <h1
      class="text-3xl font-bold mb-4"
    >
      Music Library
    </h1>
    <input
      class="w-full p-2 rounded bg-gray-800 text-white mb-4"
      placeholder="Search by title, artist, album..."
      type="text"
      value=""
    />
    <div
      class="flex space-x-4 mb-4"
    >
      <select
        class="p-2 bg-gray-800 rounded"
      >
        <option
          value="title"
        >
          Sort by Title
        </option>
        <option
          value="artist"
        >
          Sort by Artist
        </option>
        <option
          value="album"
        >
          Sort by Album
        </option>
      </select>
      <select
        class="p-2 bg-gray-800 rounded"
      >
        <option
          value="none"
        >
          All
        </option>
        <option
          value="artist"
        >
          Group by Artist
        </option>
        <option
          value="album"
        >
          Group by Album
        </option>
      </select>
    </div>
    <div
      class="mb-6"
    >
      <ul
        class="border-l-4 border-blue-500 pl-4"
      >
        <li
          class="p-2 bg-gray-800 mb-2 rounded"
        >
          ðŸŽµ 
          <strong>
            Billie Jean
          </strong>
           - Michael Jackson (Thriller)
        </li>
        <li
          class="p-2 bg-gray-800 mb-2 rounded"
        >
          ðŸŽµ 
          <strong>
            Bohemian Rhapsody
          </strong>
           - Queen (A Night at the Opera)
        </li>
        <li
          class="p-2 bg-gray-800 mb-2 rounded"
        >
          ðŸŽµ 
          <strong>
            Hey Jude
          </strong>
           - The Beatles (The Beatles)
        </li>
        <li
          class="p-2 bg-gray-800 mb-2 rounded"
        >
          ðŸŽµ 
          <strong>
            Hotel California
          </strong>
           - Eagles (Hotel California)
        </li>
        <li
          class="p-2 bg-gray-800 mb-2 rounded"
        >
          ðŸŽµ 
          <strong>
            Imagine
          </strong>
           - John Lennon (Imagine)
        </li>
        <li
          class="p-2 bg-gray-800 mb-2 rounded"
        >
          ðŸŽµ 
          <strong>
            Stairway to Heaven
          </strong>
           - Led Zeppelin (Led Zeppelin IV)
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
